teps/0004-task-results-in-final-tasks.md

```diff
@@ -139,12 +137,22 @@ With adding support for `Task` `Results` in finally `Tasks`, we are introducing
 `Tasks` but that does not change scheduling of finally tasks. All final tasks will still be executed in parallel after
 all non-final tasks are done.
 
-**Q. What happens to `Pipeline` when the dependent task `boskos-acquire` either failed or not executed and
+**Q. What happens to `Pipeline` when the dependent task `boskos-acquire` either failed or not executed, and
 the task result `leased-resource` is not initialized?**
 
-**A.** The finally `Task` `boskos-release` is attempted but declared failure since the task result
-`leased-resource` is not initialized. Pipeline controller logs the reason of this failure as invalid result
-reference `InvalidTaskResultReference`. `Pipeline` continues executing rest of the final tasks and exits with failure.
+**A.** The finally `Task` `boskos-release` is attempted and included in the list of`skippedTasks` since the task result
+`leased-resource` is not initialized. Pipeline controller logs the reason of this validation failure as invalid result
+reference `InvalidTaskResultReference`. `Pipeline` continues executing rest of the final tasks and exits with `completion`.
+
+The pipeline controller is adding a new field to capture this validation failure:
+
+```go
+type SkippedTask struct {
+  Name string `json:"name"`
+  ...
+  Reason string `json:"reason,omitempty"`
```

Comment:
i think @jerop was looking into some options around `Reason` as well

Reply:
was thinking that we can have the `Reason` as string alias to represent the reasons as constants, instead of using strings

```go
type SkippedTask struct {
        Name string `json:"name"`
        ....
        Reason ReasonSkipped `json:"reason"`
}

type ReasonSkipped string

const (
        WhenExpressionsFalse   ReasonSkipped = "ReasonWhenExpressionsFalse"
        ConditionsFailed       ReasonSkipped = "ReasonConditionsFailed"
        ParentTaskSkipped      ReasonSkipped = "ReasonParentTaskSkipped"
        ...
)
```

Reply:
thanks @bobcatfish and @jerop 

I like the idea of having such constants, how about we create a new constant "ReasonMissingTaskResultReferenceInFinallyTask"? how about adding description with details in the same struct or just leave it in the logs? Description in this scenario would include specific param name which had a task result reference and failed to resolve that param.

Reply:
> I like the idea of having such constants, how about we create a new constant "ReasonMissingTaskResultReferenceInFinallyTask"? 

yea, wondering if it's better to phrase the reason as the result is missing or the task dependency was not executed successfully -- we explored (all) the reasons for skipping in this [doc](https://docs.google.com/document/d/1I7KFeesXDo3toQY94lUePkbsdoy04Q41kuByrYlzhNw/edit?resourcekey=0-zYjfpSt6i3RjVDXklr4LtA#)

> how about adding description with details in the same struct or just leave it in the logs? 

i think we should add further details beyond the reason, not just leave it in the logs

> Description in this scenario would include specific param name which had a task result reference and failed to resolve that param

for the missing result reference, i think we should include the result reference as well in the description (not the param name only) because we'd need to do that when we support when expressions in finally tasks (because they use results directly, not through params)

Reply:
Thanks @jerop I have updated the proposal: `Pipeline controller logs the validation failure and that finally task is added into the list of `skippedTasks`.